src/main/java/org/zalando/nakadi/repository/db/EventTypeCache.java

```diff
@@ -20,75 +13,111 @@
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+
 public class EventTypeCache {
 
     public static final String ZKNODE_PATH = "/nakadi/event_types";
     public static final int CACHE_MAX_SIZE = 100000;
     private static final Logger LOG = LoggerFactory.getLogger(EventTypeCache.class);
-
-    private static class CachedValue {
-        private final EventType eventType;
-        private final EventTypeValidator eventTypeValidator;
-
-        public CachedValue(final EventType eventType, final EventTypeValidator eventTypeValidator) {
-            this.eventType = eventType;
-            this.eventTypeValidator = eventTypeValidator;
-        }
-
-        public EventType getEventType() {
-            return eventType;
-        }
-
-        public EventTypeValidator getEventTypeValidator() {
-            return eventTypeValidator;
-        }
-    }
-
     private final LoadingCache<String, CachedValue> eventTypeCache;
     private final PathChildrenCache cacheSync;
     private final ZooKeeperHolder zkClient;
     private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
+    private final TimelineSync timelineSync;
+    private Map<String, TimelineSync.ListenerRegistration> timelineRegistrations;
 
-    public EventTypeCache(final EventTypeRepository eventTypeRepository, final ZooKeeperHolder zkClient)
+    public EventTypeCache(final EventTypeRepository eventTypeRepository,
+                          final TimelineDbRepository timelineRepository,
+                          final ZooKeeperHolder zkClient,
+                          final TimelineSync timelineSync)
             throws Exception {
-        this(eventTypeRepository, zkClient, setupCacheSync(zkClient.get()));
+        this(eventTypeRepository, timelineRepository, zkClient, setupCacheSync(zkClient.get()), timelineSync);
     }
 
     @VisibleForTesting
-    EventTypeCache(final EventTypeRepository eventTypeRepository, final ZooKeeperHolder zkClient,
-                   final PathChildrenCache cache) {
+    EventTypeCache(final EventTypeRepository eventTypeRepository,
+                   final TimelineDbRepository timelineRepository,
+                   final ZooKeeperHolder zkClient,
+                   final PathChildrenCache cache,
+                   final TimelineSync timelineSync) {
         this.zkClient = zkClient;
-        this.eventTypeCache = setupInMemoryEventTypeCache(eventTypeRepository);
+        this.eventTypeCache = setupInMemoryEventTypeCache(eventTypeRepository, timelineRepository);
         this.cacheSync = cache;
+        this.timelineSync = timelineSync;
+        this.timelineRegistrations = new ConcurrentHashMap<>();
         if (null != cacheSync) {
             this.cacheSync.getListenable().addListener((curator, event) -> this.onZkEvent(event));
         }
-        preloadEventTypes(eventTypeRepository);
+        preloadEventTypes(eventTypeRepository, timelineRepository);
+    }
+
+    @PreDestroy
```

Comment:
If you acquire resource you have to clean it up when you will finish using it. i do not think there is much sense in doing it here. I will remove it if it is mess for you.